src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -0,0 +1,374 @@
+
+using System.Collections.Generic;
+using System.IO;
+using System.Net.Http.Headers;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace System.Net.Http
+{
+    internal partial class AuthenticationHelper
+    {
+        public const string Digest = "Digest";
+
+        // Define digest constants
+        private const string Qop = "qop";
+        private const string Auth = "auth";
+        private const string AuthInt = "auth-int";
+        private const string Nonce = "nonce";
+        private const string NC = "nc";
+        private const string Realm = "realm";
+        private const string UserHash = "userhash";
+        private const string Username = "username";
+        private const string Algorithm = "algorithm";
+        private const string Uri = "uri";
+        private const string Sha256 = "SHA-256";
+        private const string Md5 = "MD5";
+        private const string Sha256Sess = "SHA-256-sess";
+        private const string CNonce = "cnonce";
+        private const string Opaque = "opaque";
+        private const string Response = "response";
+        private const string Colon = ":";
+
+        // Define alphanumeric characters for cnonce
+        // 48='0', 65='A', 97='a'
+        private static int[] AlphaNumChooser = new int[] { 48, 65, 97 };
+
+        public static bool TrySetDigestAuthToken(HttpRequestMessage request, ICredentials credentials, DigestResponse digestResponse)
+        {
+            NetworkCredential credential = credentials.GetCredential(request.RequestUri, Digest);
+            if (credential == null)
+            {
+                return false;
+            }
+
+            try
+            {
+                request.Headers.Authorization = new AuthenticationHeaderValue(Digest, GetDigestTokenForCredential(credential, request, digestResponse));
+                return true;
+            }
+            catch
+            {
+                // Return false in case of any digest header calculation errors.
+                return false;
+            }
+        }
+
+        public static string GetDigestTokenForCredential(NetworkCredential credential, HttpRequestMessage request, DigestResponse digestResponse)
+        {
+            StringBuilder sb = StringBuilderCache.Acquire();
+
+            // It is mandatory for servers to implement sha-256
+            // Keep MD5 for backward compatibility.
+            string algorithm = Md5;
+            if (digestResponse.Parameters[Algorithm].Contains(Sha256))
+            {
+                algorithm = Sha256;
+            }
+
+            string realm = digestResponse.Parameters.ContainsKey(Realm) ? digestResponse.Parameters[Realm] : string.Empty;
+
+            // Add username
+            if (digestResponse.Parameters.ContainsKey(UserHash) && digestResponse.Parameters[UserHash] == "true")
+            {
+                sb.AppendKeyValue(Username, ComputeHash(digestResponse.Parameters[Username] + Colon + realm, algorithm));
```

Comment:
This might be what's breaking connecting to the test server!! If it still doesn't work after this, can you post the servers challenge and the generated authentication response.